README.md
# Shoppings

[![Build Status](https://www.travis-ci.com/gissandrogama/-shopping.svg?branch=main)](https://www.travis-ci.com/gissandrogama/-shopping)

[![codecov](https://codecov.io/gh/gissandrogama/-shopping/branch/main/graph/badge.svg?token=4Z7ZWDKX5V)](https://codecov.io/gh/gissandrogama/-shopping)

technical test for Stone's training program

# Technical information
  * Elixir Ecosystem
  * credo to help keep the code clean and with quality
  * excoveralls help maintain and increase application test coverage
  * mix_test_watch facilitate the correction of tests and help to follow a TDD practice
  * ex_doc and earmark, generate documentation

## Installation

## Iniciando a api localmente

**1.** Clone the project:

  * ssh
  ```sh
  $ git clone git@github.com:gissandrogama/-shopping.git
  ```

  * https
  ```sh
  $ git clone https://github.com/gissandrogama/-shopping
  ```

**2.** Access the project folder:

```sh
$ cd -shopping
```

**3.** Install the dependencies:

```sh
$ mix deps.get
```

**5.** start the application:

```sh
$ iex -S mix
```

in item values ​​was from the shopping list. They are whole numbers, to avoid arithmetic complexity with floating point or float. Thus, to represent R $ 1.99, 199 or 1_99 is used, for R $ 50.00 -> 50_00 or 5000.

**5.** generate documentation:
In the documentation has explaining how to use the application.

```sh
$ mix docs
```





CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased
- refactor

## [0.0.1] - 2021-01-18
### Added

- project start

- credo configuration

- coverralls configuration

- creating license files

- creation of the buy function

- creation of the handle module and functions: calculate, split and process

- written documentation to facilitate understanding and how to use the application

- added ex_doc and earmark libs to generate documentation

### Changed

- the readme has been rewritten with instructions to start the application
- the calculate, split and process functions have been changed to handle floating-point or float arithmetic



test/handle_test.exs
defmodule HandleTest do
  use ExUnit.Case
  doctest Handle

  alias Handle

  describe "calculate/2" do
    test "return empty list" do
      itens = []
      emails = []
      result = Handle.calculate(itens, emails)
      assert result == []
    end

    test "when an item of 1 real is divided for 3 people" do
      itens = [%{name: "chocolate stick", amount: 1, price: 1_00, type: "unity"}]
      emails = ["teste1@gmail.com", "teste2@gmail.com", "teste3@gmail.com"]
      result = Handle.calculate(itens, emails)

      assert result == [
               %{email: "teste1@gmail.com", value: "R$ 0.33"},
               %{email: "teste2@gmail.com", value: "R$ 0.33"},
               %{email: "teste3@gmail.com", value: "R$ 0.34"}
             ]
    end

    test "several items for several people" do
      itens = [
        %{name: "contra file", amount: 3, price: 5000, type: "kg"},
        %{name: "leite liquido", amount: 4, price: 399, type: "litro"},
        %{name: "arroz", amount: 4, price: 598, type: "kg"},
        %{name: "feijão", amount: 4, price: 799, type: "kg"},
        %{name: "queijo mussarela", amount: 4, price: 4289, type: "kg"}
      ]

      emails = ["teste1@gmail.com", "teste2@gmail.com", "teste3@gmail.com", "teste4@gmail.com"]
      result = Handle.calculate(itens, emails)

      assert result == [
               %{email: "teste1@gmail.com", value: "R$ 98.35"},
               %{email: "teste2@gmail.com", value: "R$ 98.35"},
               %{email: "teste3@gmail.com", value: "R$ 98.35"},
               %{email: "teste4@gmail.com", value: "R$ 98.35"}
             ]
    end

    test "several items for one person" do
      itens = [
        %{name: "contra file", amount: 3, price: 5000, type: "kg"},
        %{name: "leite liquido", amount: 4, price: 399, type: "litro"},
        %{name: "arroz", amount: 4, price: 598, type: "kg"},
        %{name: "feijão", amount: 4, price: 799, type: "kg"},
        %{name: "queijo mussarela", amount: 4, price: 4289, type: "kg"},
        %{name: "queijo mussarela", amount: 4, price: 4289, type: "kg"}
      ]

      emails = ["teste1@gmail.com"]
      result = Handle.calculate(itens, emails)

      assert result == [%{email: "teste1@gmail.com", value: "R$ 564.96"}]
    end

    test "many items for many people" do
      itens = [
        %{name: "contra file", amount: 3, price: 5000, type: "kg"},
        %{name: "leite liquido", amount: 4, price: 399, type: "litro"},
        %{name: "arroz", amount: 4, price: 598, type: "kg"},
        %{name: "feijão", amount: 4, price: 799, type: "kg"},
        %{name: "queijo mussarela", amount: 4, price: 4289, type: "kg"},
        %{name: "queijo mussarela", amount: 4, price: 4289, type: "kg"},
        %{name: "contra file", amount: 3, price: 5000, type: "kg"},
        %{name: "leite liquido", amount: 4, price: 399, type: "litro"},
        %{name: "arroz", amount: 4, price: 598, type: "kg"},
        %{name: "feijão", amount: 4, price: 799, type: "kg"},
        %{name: "queijo mussarela", amount: 4, price: 4289, type: "kg"},
        %{name: "queijo mussarela", amount: 4, price: 4289, type: "kg"},
        %{name: "contra file", amount: 3, price: 5000, type: "kg"},
        %{name: "leite liquido", amount: 4, price: 399, type: "litro"},
        %{name: "arroz", amount: 4, price: 598, type: "kg"},
        %{name: "feijão", amount: 4, price: 799, type: "kg"},
        %{name: "queijo mussarela", amount: 4, price: 4289, type: "kg"},
        %{name: "queijo mussarela", amount: 4, price: 4289, type: "kg"}
      ]

      emails = [
        "teste1@gmail.com",
        "teste2@gmail.com",
        "teste3@gmail.com",
        "teste4@gmail.com",
        "teste5@gmail.com",
        "teste6@gmail.com",
        "teste7@gmail.com",
        "teste8@gmail.com",
        "teste9@gmail.com"
      ]

      result = Handle.calculate(itens, emails)

      assert result == [
               %{email: "teste1@gmail.com", value: "R$ 188.32"},
               %{email: "teste2@gmail.com", value: "R$ 188.32"},
               %{email: "teste3@gmail.com", value: "R$ 188.32"},
               %{email: "teste4@gmail.com", value: "R$ 188.32"},
               %{email: "teste5@gmail.com", value: "R$ 188.32"},
               %{email: "teste6@gmail.com", value: "R$ 188.32"},
               %{email: "teste7@gmail.com", value: "R$ 188.32"},
               %{email: "teste8@gmail.com", value: "R$ 188.32"},
               %{email: "teste9@gmail.com", value: "R$ 188.32"}
             ]
    end
  end

  describe "split/3" do
    test "when it is fraction and infinite tithe" do
      result =
        Handle.split(89_53, 3, ["teste1@gmail.com", "teste2@gmail.com", "teste3@gmail.com"])

      assert result == [
               %{email: "teste1@gmail.com", value: "R$ 29.84"},
               %{email: "teste2@gmail.com", value: "R$ 29.84"},
               %{email: "teste3@gmail.com", value: "R$ 29.85"}
             ]
    end

    test "when it is integer" do
      result = Handle.split(5000, 2, ["teste1@gmail.com", "teste2@gmail.com"])

      assert result == [
               %{email: "teste1@gmail.com", value: "R$ 25.0"},
               %{email: "teste2@gmail.com", value: "R$ 25.0"}
             ]
    end
  end
end


test/shoppings_test.exs
defmodule ShoppingsTest do
  use ExUnit.Case
  doctest Shoppings

  describe "buy/0" do
    test "taking values ​​from the macros defined in the Shopping module" do
      result = Shoppings.buy()

      assert result == [
               %{email: "teste1@gmail.com", value: "R$ 0.33"},
               %{email: "teste2@gmail.com", value: "R$ 0.33"},
               %{email: "teste3@gmail.com", value: "R$ 0.34"}
             ]
    end
  end

  describe "buy/2" do
    test "passing shopping list and emails" do
      buys = [
        %{name: "baton garoto", amount: 1, price: 1_00, type: "unity"},
        %{name: "contra file", amount: 1, price: 50_00, type: "kg"}
      ]

      emails = ["teste1@gmail.com", "teste2@gmail.com", "teste3@gmail.com"]

      result = Shoppings.buy(buys, emails)

      assert result == [
               %{email: "teste1@gmail.com", value: "R$ 17.0"},
               %{email: "teste2@gmail.com", value: "R$ 17.0"},
               %{email: "teste3@gmail.com", value: "R$ 17.0"}
             ]
    end
  end
end


mix.exs
defmodule Shoppings.MixProject do
  use Mix.Project

  @github_url "https://github.com/gissandrogama/-shopping.git"

  def project do
    [
      app: :shoppings,
      version: "0.1.0",
      elixir: "~> 1.11",
      description: "Stone's technical deficit for the Elixir Training Program",
      source_url: @github_url,
      homepage_url: @github_url,
      files: ~w(mix.exs lib LICENSE.md README.md CHANGELOG.md),
      package: [
        maintainers: ["Gissandro Gama"],
        licenses: ["MIT"],
        links: %{
          "Github" => @github_url
        }
      ],
      docs: [
        main: "readme",
        extras: ["README.md", "CHANGELOG.md"]
      ],
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      test_coverage: [tool: ExCoveralls],
      preferred_cli_env: [
        coveralls: :test,
        "coveralls.detail": :test,
        "coveralls.post": :test,
        "coveralls.json": :test,
        "coveralls.html": :test
      ]
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:credo, "~> 1.5", only: [:dev, :test], runtime: false},
      {:excoveralls, "~> 0.13.4", only: :test},
      {:mix_test_watch, "~> 1.0", only: :dev, runtime: false},
      {:ex_doc, "~> 0.23", only: :dev, runtime: false},
      {:earmark, "~> 1.3", only: [:dev]}
      # {:dep_from_hexpm, "~> 0.3.0"},
      # {:dep_from_git, git: "https://github.com/elixir-lang/my_dep.git", tag: "0.1.0"}
    ]
  end
end


lib/handle.ex
defmodule Handle do
  @moduledoc """
  Handle module for list entries such as `items` and `emails`.
  The most used function and the `calculate/2` function.
  """

  @doc """
  Function calculates the total value of a list of items,
  divides the total amount according to the amount and past emails.

  ## Function parameters

    - items: list of maps contains information about the list item
    - emails: list of people who will be divided the total amount

  ## Additional Information

    - if the parameters passed are empty lists, the function returns an empty list

  ## Example empty lists

       iex> Handle.calculate([], [])
       []

  ## Example non-empty lists

      iex> Handle.calculate([%{name: "arroz", amount: 4, price: 598, type: "kg"}, %{name: "feijão", amount: 4, price: 799, type: "kg"}, %{name: "queijo mussarela", amount: 4, price: 4289, type: "kg"}], ["teste1@gmail.com", "teste2@gmail.com", "teste3@gmail.com", "teste4@gmail.com"])
      [
         %{email: "teste1@gmail.com", value: "R$ 56.86"},
         %{email: "teste2@gmail.com", value: "R$ 56.86"},
         %{email: "teste3@gmail.com", value: "R$ 56.86"},
         %{email: "teste4@gmail.com", value: "R$ 56.86"}
      ]
  """
  def calculate([], []), do: []

  @spec calculate(list(), list()) :: [map()]
  def calculate(items, emails) do
    amount = items |> Enum.map(fn item -> item.amount * item.price end) |> Enum.sum()
    people = emails |> Enum.count()

    split(amount, people, emails)
  end

  @doc """
  Function calculates apportionment of how much each person must pay.
  And it displays in a `map` with the email of each person and the amount of must be paid.

  ## Function parameters

    - value: total value of the list of items, the result of the sum of the quantity
            and the unique value of each item.
    - quantity: number of people / emails from the email list
    - emails: the email list

  ## Additional Information

    - function returns a `list` of` map`:
    ```Elixir
      [%{email: teste@email.com, value: "R$ 50.00"}]
    ```

  ## Example

       iex> Handle.split(100, 3, ["teste1@gmail.com", "teste2@gmail.com", "teste3@gmail.com"])
       [
        %{email: "teste1@gmail.com", value: "R$ 0.33"},
        %{email: "teste2@gmail.com", value: "R$ 0.33"},
        %{email: "teste3@gmail.com", value: "R$ 0.34"}
       ]
  """
  @spec split(number, number, list()) :: [map()]
  def split(value, quantity, emails) do
    cloven = value / quantity

    case rem(value, quantity) do
      0 ->
        value = cloven

        Enum.map(emails, fn email -> %{email: email, value: value} end) |> print()

      _ ->
        value = process(cloven) |> String.to_integer()

        list = Enum.map(emails, fn email -> %{email: email, value: value} end)

        key = Enum.count(list) - 1

        list
        |> List.update_at(key, fn pessoas ->
          %{email: pessoas.email, value: pessoas.value + 1}
        end)
        |> print()
    end
  end

  @doc """
  the function receives an infinite decimal decimal or a fractional decimal. Put it to two decimal places
  and return an instring in integer format.

  ## Function parameters

    - clove: value obtained by dividing the sum of items by number of people / email

  ## Additional Information

    - this function only reaches numbers with infinite decimals or fractions with more than 3 decimal places

  ## Example

       iex> Handle.process(33.33333333333)
       "033"

       iex> Handle.process(2986.333333333336)
       "2986"
  """
  @spec process(float) :: String.t()
  def process(cloven) when is_float(cloven) do
    cloven = cloven / 100

    cloven
    |> Float.floor(2)
    |> Float.to_string()
    |> String.split(".")
    |> Enum.join()
  end

  defp print(list) do
    Enum.map(list, fn list -> %{email: list.email, value: "R$ #{list.value / 100}"} end)
  end
end


lib/shoppings.ex
defmodule Shoppings do
  @moduledoc """
  module has function that receives lists
  """

  @doc """
  function to receive a shopping list and an email list. She passes on these lists to the `calculate / 2` function of the` Handle` module, to perform the due processing.

  ## Function parameters

    - buy: list of maps contains information items of a purchase
    - emails: list of emails that represent people who will share the total purchase price

  ## Additional Information

    - if the parameters passed are empty lists, the function returns an empty list

  ## Example buy/0

       iex> Shoppings.buy()
       [
         %{email: "teste1@gmail.com", value: "R$ 0.33"},
         %{email: "teste2@gmail.com", value: "R$ 0.33"},
         %{email: "teste3@gmail.com", value: "R$ 0.34"}
       ]

  ## Example buy/2

      iex> Shoppings.buy([%{name: "arroz", amount: 4, price: 598, type: "kg"}, %{name: "feijão", amount: 4, price: 799, type: "kg"}, %{name: "queijo mussarela", amount: 4, price: 4289, type: "kg"}], ["teste1@gmail.com", "teste2@gmail.com", "teste3@gmail.com", "teste4@gmail.com"])
      [
         %{email: "teste1@gmail.com", value: "R$ 56.86"},
         %{email: "teste2@gmail.com", value: "R$ 56.86"},
         %{email: "teste3@gmail.com", value: "R$ 56.86"},
         %{email: "teste4@gmail.com", value: "R$ 56.86"}
      ]
  """
  @buys [%{name: "baton garoto", amount: 1, price: 100, type: "unity"}]
  @emails ["teste1@gmail.com", "teste2@gmail.com", "teste3@gmail.com"]
  def buy do
    Handle.calculate(@buys, @emails)
  end

  def buy(buys, emails) do
    Handle.calculate(buys, emails)
  end
end
